### PR TITLE
Fix filter check errors for special weapons

### PR DESCRIPTION
**Pull request by newfrenchy83**
This is to repair two errors introduced in the conditions of definition of tag_name variables, used to determine if an attack variable can be checked taking into account or not the modifications introduced by a special weapon.

First error, when apply_to=both, the distinction between AFFECT_SELF/OTHER must not be made to avoid an infinite loop.

2nd error, for the check of [filter_attacker/defender] the check of the variable is_attacker must be associated with AFFECT_SELF/OTHER in the following way: 
If special affects self then is_attacker is true if [filter_attacker] and false if [filter_defender].

If special affects opponent then it's the opposite.

### CODE REVIEW

**Comment by newfrenchy83** [comment]
@stevecotton could you review this PR before https://github.com/wesnoth/wesnoth/pull/8977/files#diff-eb9c8f8b8122951ad515f39be0e4d8550c4f22421024154f55bdef508de85897 please?

**Comment by newfrenchy83** [comment]
@Pentarctagon @stevecotton could you review and merge this PR, please?

**Comment by stevecotton** [comment]
Please add unit tests

**Comment by newfrenchy83** [comment]
> Please add unit tests

done

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_attacker_defender_filtering.cfg [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#textdomain wesnoth-test
+
+#####
+# API(s) being tested: aplpy_to=both with [filter_attacker][filter_weapon]type= and [filter_defender][filter_weapon]type=
```
Comment (Pentarctagon) [inline data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_attacker_defender_filtering.cfg:4]: `apply` is misspelled.

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_applied_to_both.cfg [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#textdomain wesnoth-test
+
+#####
+# API(s) being tested: aplpy_to=both with [filter_self][filter_weapon]type= and [filter_opponent][filter_weapon]type=
```
Comment (Pentarctagon) [inline data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_applied_to_both.cfg:4]: `apply` is misspelled.

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_attacker_defender_filtering.cfg [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+##
+# Actions:
+# Alice and Bob are both of type Test Melee Quintain.
+# chane type of attack of pierce and resistance to -50%.
```
Comment (Pentarctagon) [inline data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_attacker_defender_filtering.cfg:8]: `change` is misspelled.

**Comment by Pentarctagon** [comment]
> First error, when apply_to=both, the distinction between AFFECT_SELF/OTHER must not be made to avoid an infinite loop.

What situation would lead to that causing an infinite loop?

> 2nd error, for the check of [filter_attacker/defender] the check of the variable is_attacker must be associated with 
> AFFECT_SELF/OTHER in the following way:
> If special affects self then is_attacker is true if [filter_attacker] and false if [filter_defender].
> 
> If special affects opponent then it's the opposite.

Why would `is_attacker` change depending on whether `affect_self` vs `affect_enemeies`/`affect_allies` is used? What happens if something uses both `affect_self` and `affect_allies`, for example?

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_attacker_defender_filtering.cfg (hunk 3 above)
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_attacker_defender_filtering.cfg]: fixed

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_applied_to_both.cfg [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+##
+# Actions:
+# Alice and Bob are both of type Test Melee Quintain.
+# chane type of attack of pierce and resistance to -50%.
```
[the remark this answers is not in the document]
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_applied_to_both.cfg]: fixed

File: data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_applied_to_both.cfg (hunk 2 above)
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/damage_type/damage_type_applied_to_both.cfg]: fixed

**Comment by newfrenchy83** [comment]
> > First error, when apply_to=both, the distinction between AFFECT_SELF/OTHER must not be made to avoid an infinite loop.
> 
> What situation would lead to that causing an infinite loop?
> 
> > 2nd error, for the check of [filter_attacker/defender] the check of the variable is_attacker must be associated with
> > AFFECT_SELF/OTHER in the following way:
> > If special affects self then is_attacker is true if [filter_attacker] and false if [filter_defender].
> > If special affects opponent then it's the opposite.
> 
> Why would `is_attacker` change depending on whether `affect_self` vs `affect_enemeies`/`affect_allies` is used? What happens if something uses both `affect_self` and `affect_allies`, for example?

if i use in [damage_type] both [filter_attacker]type= and [filter_defender]type= with apply_to=attacker|defender i has recursion issue if only is_attacker variable used. for apply_to=both when used tag_name must be pushed in all filter for what original type only be checked.

**Comment by Pentarctagon** [comment]
That doesn't really answer my question - I'm asking *why* using only `is_attacker` causes an infinite loop.

**Comment by newfrenchy83** [comment]
> That doesn't really answer my question - I'm asking _why_ using only `is_attacker` causes an infinite loop.

ithink what it because usinf of AFFECT_EITHER in pre attack windows who cause the issue but i'm not sure myself

**Comment by Pentarctagon** [comment]
You should know what's causing a problem before implementing a fix for that problem.

**Comment by newfrenchy83** [comment]
> You should know what's causing a problem before implementing a fix for that problem.

when I reread the function special_affects_opponent(const config& special, bool is_attacker), I notice that the fiction matches if is_attacker is true with apply_to=defender when whom=AFFECT_OTHER or, when I use [filter_attacker]type= with apply_to=defender, while the defender is not the owner of the weapon, I get a recursion, because in this case is_attacker must be false to designate the attacker and true when the defender is the owner of the weapon; it is therefore necessary to specify when is_attacker=true is valid to designate the attacker following that whom=AFFECT_SELF or OTHER

**Comment by newfrenchy83** [comment]
i rewrite documentation in abilities.cpp

**Comment by Pentarctagon** [comment]
Thinking about it some more, something that I think would help me given the complexity of the abilities code would be to have failing unit tests added in one PR -> merge that PR -> then have the fix in a separate PR. That way I'd be able to much more easily reproduce and debug through the issue locally as I'm trying to understand the fix.

@stevecotton does that approach make sense?

**Comment by newfrenchy83** [comment]
> Thinking about it some more, something that I think would help me given the complexity of the abilities code would be to have failing unit tests added in one PR -> merge that PR -> then have the fix in a separate PR. That way I'd be able to much more easily reproduce and debug through the issue locally as I'm trying to understand the fix.
> 
> @stevecotton does that approach make sense?

I will change the test of thisPr and add in séparante Pr for show bug

**Comment by newfrenchy83** [comment]
> Thinking about it some more, something that I think would help me given the complexity of the abilities code would be to have failing unit tests added in one PR -> merge that PR -> then have the fix in a separate PR. That way I'd be able to much more easily reproduce and debug through the issue locally as I'm trying to understand the fix.
> 
> @stevecotton does that approach make sense?

test added in https://github.com/wesnoth/wesnoth/pull/9251 PR

**Comment by soliton-** [comment]
> Thinking about it some more, something that I think would help me given the complexity of the abilities code would be to have failing unit tests added in one PR -> merge that PR -> then have the fix in a separate PR. That way I'd be able to much more easily reproduce and debug through the issue locally as I'm trying to understand the fix.

Assuming that the tests are marked as expected to fail at first that sounds perfectly fine.

**Comment by newfrenchy83** [comment]
@Pentarctagon @soliton- PR rebased and ready to review

**Comment by Pentarctagon** [comment]
So for the changes made here, they're for determining the value of the last argument provided to `special_unit_matches()`. Inside `special_unit_matches()`, the `tag_name` argument is only passed to the weapon's `matches_filter()` argument. From there is passes `tag_name` to the `matches_simple_filter()` method, which only uses `tag_name` to check whether the tag is `[damage_type]` or not.

If the tag is `[damage_type]` and it's filtering on the attack's damage type, then it uses the attack's base type (ignoring `[damage_type]`). If the tag isn't `[damage_type]`, then it takes any `[damage_type]` effects into consideration.

So this ends up being, for `[damage_type]`:
* When checking `special_unit_matches()` for `self`, pass in `"damage_type"` instead of an empty string when `apply_to=both`.
* When checking `special_unit_matches()` for `other`, pass in `"damage_type"` instead of an empty string when `apply_to=both`.
* When checking `special_unit_matches()` for `att` (the attacker), pass in `"damage_type"` instead of an empty string when `apply_to=both`, when this unit is the attacker and this special affects the attacker, or when this unit is not the attacker and this special doesn't affect the attacker.
* When checking `special_unit_matches()` for `def` (the defender), pass in `"damage_type"` instead of an empty string when `apply_to=both`, when this unit is the defender and this special affects the defender, or when this unit is not the defender and this special doesn't affect the defender.

This fixes the recursion issue because it checks against the base damage type rather than getting stuck in a loop evaluating what [damage_type] should do.

**Comment by newfrenchy83** [comment]
> So for the changes made here, they're for determining the value of the last argument provided to `special_unit_matches()`. Inside `special_unit_matches()`, the `tag_name` argument is only passed to the weapon's `matches_filter()` argument. From there is passes `tag_name` to the `matches_simple_filter()` method, which only uses `tag_name` to check whether the tag is `[damage_type]` or not.
> 
> If the tag is `[damage_type]` and it's filtering on the attack's damage type, then it uses the attack's base type (ignoring `[damage_type]`). If the tag isn't `[damage_type]`, then it takes any `[damage_type]` effects into consideration.
> 
> So this ends up being, for `[damage_type]`:
> 
>     * When checking `special_unit_matches()` for `self`, pass in `"damage_type"` instead of an empty string when `apply_to=both`.
> 
>     * When checking `special_unit_matches()` for `other`, pass in `"damage_type"` instead of an empty string when `apply_to=both`.
> 
>     * When checking `special_unit_matches()` for `att` (the attacker), pass in `"damage_type"` instead of an empty string when `apply_to=both`, when this unit is the attacker and this special affects the attacker, or when this unit is not the attacker and this special doesn't affect the attacker.
> 
>     * When checking `special_unit_matches()` for `def` (the defender), pass in `"damage_type"` instead of an empty string when `apply_to=both`, when this unit is the defender and this special affects the defender, or when this unit is not the defender and this special doesn't affect the defender.
> 
> 
> This fixes the recursion issue because it checks against the base damage type rather than getting stuck in a loop evaluating what [damage_type] should do.

exact.

**Comment by newfrenchy83** [comment]
@Pentarctagon have added somemodifications in https://github.com/wesnoth/wesnoth/pull/9119/commits/991a099276a92c3b354ecc60be4154b0f4e16bab. Could you review this before to merge  this PR?

**Comment by soliton-** [comment]
> * When checking `special_unit_matches()` for `att` (the attacker), pass in `"damage_type"` instead of an empty string when `apply_to=both`, when this unit is the attacker and this special affects the attacker, or when this unit is not the attacker and this special doesn't affect the attacker.

The last case basically means the other unit is the attacker and the special affects the other unit, right?

---

It would be good to clarify (with a better name or comments) that `tag_name` is not just the tag name but the tag name if relevant for recursion or empty.

**Comment by newfrenchy83** [comment]
> > * When checking `special_unit_matches()` for `att` (the attacker), pass in `"damage_type"` instead of an empty string when `apply_to=both`, when this unit is the attacker and this special affects the attacker, or when this unit is not the attacker and this special doesn't affect the attacker.
> 
> The last case basically means the other unit is the attacker and the special affects the other unit, right?
> 
> It would be good to clarify (with a better name or comments) that `tag_name` is not just the tag name but the tag name if relevant for recursion or empty.

@soliton- check_special_is_callable is a good idea or have you another suggestion?

**Comment by newfrenchy83** [comment]
@soliton- check_if_recursion is perhaps more explicit.

**Comment by Pentarctagon** [comment]
@soliton- any additional concerns?

**Comment by soliton-** [comment]
The second commit does not look sensible to me.

In general I don't know what the plan with recursion handling is. As far as I know there is a general recursion limit now and also these tag name based checks. We want to keep these in parallel?

Since I guess the first commit is just fixing deficiencies in the tag name based checks I suppose it can go in either way but it'd be good to know what the general approach is.

Besides that the commit messages and code comments would need to be improved. I'm happy to do that though I'd prefer to rewrite the commits in one go and force push instead of a lot of back and forth commenting.

**Comment by newfrenchy83** [comment]
> The second commit does not look sensible to me.
> 
> In general I don't know what the plan with recursion handling is. As far as I know there is a general recursion limit now and also these tag name based checks. We want to keep these in parallel?
> 
> Since I guess the first commit is just fixing deficiencies in the tag name based checks I suppose it can go in either way but it'd be good to know what the general approach is.
> 
> Besides that the commit messages and code comments would need to be improved. I'm happy to do that though I'd prefer to rewrite the commits in one go and force push instead of a lot of back and forth commenting.

merged commit and rewrite comments.

**Comment by Pentarctagon** [comment]
Is it a valid case to have two abilities of the same type, where one of them filters on the other one?

**Comment by newfrenchy83** [comment]
> Is it a valid case to have two abilities of the same type, where one of them filters on the other one?

if you are talking about using special_id_active= the tag_name check does not concern it to take the example of [damage_type] since it is the only type of special weapon currently concerned; the tag_name check only concerns the filtering of the damage type, if two damage_type applied to self use [filter_self][filter_weapon]type=, only the unmodified damage type will be detectable, but if one of the two is active if the other is also active, it should be able to check it normally if i understand your question.

**Comment by Pentarctagon** [comment]
@stevecotton any concerns about the change to the recursion handling here?

**Comment by soliton-** [comment]
> > The second commit does not look sensible to me.
> > In general I don't know what the plan with recursion handling is. As far as I know there is a general recursion limit now and also these tag name based checks. We want to keep these in parallel?
> > Since I guess the first commit is just fixing deficiencies in the tag name based checks I suppose it can go in either way but it'd be good to know what the general approach is.
> > Besides that the commit messages and code comments would need to be improved. I'm happy to do that though I'd prefer to rewrite the commits in one go and force push instead of a lot of back and forth commenting.
> 
> merged commit and rewrite comments.

Did you just merge the second commit into the first? I said the second commit does not look sensible to me. I do not agree that it should be merged. So now I do not agree this should be merged at all. (I assume this is a communication error or I would find this behavior very rude.)

**Comment by newfrenchy83** [comment]
> > > The second commit does not look sensible to me.
> > > In general I don't know what the plan with recursion handling is. As far as I know there is a general recursion limit now and also these tag name based checks. We want to keep these in parallel?
> > > Since I guess the first commit is just fixing deficiencies in the tag name based checks I suppose it can go in either way but it'd be good to know what the general approach is.
> > > Besides that the commit messages and code comments would need to be improved. I'm happy to do that though I'd prefer to rewrite the commits in one go and force push instead of a lot of back and forth commenting.
> > 
> > 
> > merged commit and rewrite comments.
> 
> Did you just merge the second commit into the first? I said the second commit does not look sensible to me. I do not agree that it should be merged. So now I do not agree this should be merged at all. (I assume this is a communication error or I would find this behavior very rude.)

@soliton- i removed second commit content of this PR, excuses me for my misunderstanding.

**Review by soliton- — APPROVED** [review]
Having the `tag_name` -> `check_if_recursion` renaming in its own commit was better but I'm fine with this PR either way.

**Comment by Pentarctagon** [comment]
I will go ahead and merge this tomorrow then.

**Pull request merged**